### PR TITLE
NF: Adding N4 bias correction deep learning model

### PR DESCRIPTION
**Pull request by praitayini**
Performs bias field correction

Main features added:
nn/deepn4.py - Resamples T1 input data, loads in deepn4 weights, predicts the bias field and corrects the T1 image

Additional requirements
nn/tests/test_deepn4.py - Test script for nn/deepn4.py
nn/meson.build, nn/tests/meson.build - Updated with deepn4 files
data/fetcher.py - Updated with figshare links for weights and test input / output
workflows/tests/test_io.py - Updated datanames

### CODE REVIEW

**Comment by pjsjongsung** [comment]
Thank you for the PR! Could you
1. match pep8 style as suggested above
2. merge the commit into one and change the commit message(and this PR's description) so that it would have a tag and more detailed description? (Something like NF: Adding N4 bias correction deep learning model)

**Comment by pjsjongsung** [comment]
Thank you! There seems to be just one pep8 left, if you could resolve it as well.
I am yet to test it on any images, but the code looks good.

File: dipy/nn/deepn4.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+from dipy.nn.utils import transform_img, recover_img
+
+from dipy.data import get_fnames
+from dipy.testing.decorators import doctest_skip_parser
+from dipy.utils.optpkg import optional_package
+from dipy.nn.utils import set_logger_level
+
+tf, have_tf, _ = optional_package('tensorflow')
+tfa, have_tfa, _ = optional_package('tensorflow_addons')
+if have_tf and have_tfa:
```
Comment (pjsjongsung) [inline dipy/nn/deepn4.py:19]: Can you change this part like in Synb0? I add the Layer and Model as None because otherwise without tensorflow importing this throws an error.

File: dipy/nn/deepn4.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    from tensorflow.keras.models import Model
+    from tensorflow.keras.layers import MaxPool3D, Conv3DTranspose
+    from tensorflow.keras.layers import Conv3D, LeakyReLU
+    from tensorflow.keras.layers import Concatenate, Layer
+    from tensorflow_addons.layers import InstanceNormalization
+else:
+    logging.warning('This model requires Tensorflow.\
```
Comment (pjsjongsung) [inline dipy/nn/deepn4.py:26]: Also mention Tensorflow add-ons, so like the Synb0 message

File: dipy/nn/deepn4.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+                             do not match the declared model')
+
+    def __predict(self, x_test):
+        r"""
+        Internal prediction function
+        Predict bias field from input T1 signal
+
+        Parameters
+        ----------
+        x_test : np.ndarray (1, 1, 128, 128, 128)
```
Comment (pjsjongsung) [inline dipy/nn/deepn4.py:225]: Channel is last dimension.

File: dipy/nn/deepn4.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    This class is intended for the DeepN4 model.
+    """
+
+    @doctest_skip_parser
+    def __init__(self, verbose=False):
+        r"""
+
+        To obtain the pre-trained model, use::
+        >>> deepn4_model = DeepN4() # skip if not have_tf
+        >>> fetch_model_weights_path = get_fnames('model_weights')
```
Comment (pjsjongsung) [inline dipy/nn/deepn4.py:156]: I think the example should be just using fetch_default_weights function?

Comment (praitayini) [reply dipy/nn/deepn4.py]: I changed this to 
 ```
To obtain the pre-trained model, use fetch_default_weights() like:
        >>> deepn4_model = DeepN4()
        >>> fetch_model_weights_path = get_fnames('deepn4_default_weights')
        >>> deepn4_model.load_model_weights(fetch_model_weights_path)
```
let me know if this is good!

File: dipy/nn/deepn4.py (hunk 3 above)
Comment (praitayini) [reply dipy/nn/deepn4.py]: changed it to `x_test : np.ndarray (128, 128, 128, 1)`

File: dipy/nn/deepn4.py (hunk 2 above)
Comment (praitayini) [reply dipy/nn/deepn4.py]: updated to 
```
logging.warning('This model requires Tensorflow and Tensorflow\
                    -addons. Please install these packages using \
                    pip. If using mac, please refer to this \
                    link for installation. \
                    https://github.com/apple/tensorflow_macos')
```

File: dipy/nn/deepn4.py (hunk 1 above)
Comment (praitayini) [reply dipy/nn/deepn4.py:19]: added     ```
from tensorflow_addons.layers import InstanceNormalization
else:
    class Model:
        pass

    class Layer:
        pass
    logging.warning('This model requires Tensorflow and Tensorflow\
```

**Comment by skoudoro** [comment]
Hi @praitayini ,

We fixed the issue in master,  Can you rebase this branch or merge master to this branch?

Thanks

File: dipy/nn/deepn4.py (hunk 4 above)
Comment (pjsjongsung) [reply dipy/nn/deepn4.py]: The example description talks about using fetch_default_weights() but the actual example does not use fetch_default_weights function.

File: dipy/nn/deepn4.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        tmp[lx:lX, ly:lY, lz:lZ] = img[rx:rX, ry:rY, rz:rZ]
+
+        return tmp, [lx, lX, ly, lY, lz, lZ, rx, rX, ry, rY, rz, rZ]
+
+    def normalize_img(self, img, max_img, min_img, a_max, a_min):
+
+        img = (img - min_img) / (max_img - min_img)
+        img = np.clip(img, a_max=a_max, a_min=a_min)
+
+        return img
```
Comment (skoudoro) [inline dipy/nn/deepn4.py:269]: to remove.  You can use `from dipy.nn.utils import normalize`

File: dipy/nn/deepn4.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
+#!/usr/bin/python
+"""
+Class and helper functions for fitting the DeepN4 model.
+"""
+import logging
+import numpy as np
+
+import nibabel as nib
```
Comment (skoudoro) [inline dipy/nn/deepn4.py:8]: Can you remove this, it is not used

**Review by skoudoro — COMMENTED** [review]
Hi @praitayini,

Overall, it looks good. Can you address the small comments below and then, we can go ahead and merge this PR today.  Thanks !

File: dipy/nn/deepn4.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+    """
+    This class is intended for the DeepN4 model.
+    """
+
+    @doctest_skip_parser
+    def __init__(self, verbose=False):
+        r"""
+
+        To obtain the pre-trained model, use fetch_default_weights() like:
+        >>> deepn4_model = DeepN4()
```
Comment (skoudoro) [inline dipy/nn/deepn4.py:160]: it is missing `# skip if not have_tf`

That 's why it is failing

File: dipy/nn/deepn4.py [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+        >>> deepn4_model.fetch_default_weights()
```
Comment (skoudoro) [inline dipy/nn/deepn4.py:161]: it is missing `# skip if not have_tf`

That 's why it is failing

**Comment by skoudoro** [comment]
Thank you for the update @praitayini,

Waiting for the CI's to make sure it is all green and then, merging.

**Comment by skoudoro** [comment]
Thank you @praitayini for this work.

I am going ahead and merging this PR
